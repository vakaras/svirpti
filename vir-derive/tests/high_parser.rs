use svirpti_vir_derive::vir_high;

#[test]
fn unary_expression() {
    let program = vir_high! {
        procedure {
            locals {
                a: Bool,
            }
            bb1 {
                guard !a && true;
                goto { exit }
            }
        }
    };
    insta::assert_yaml_snapshot!(program);
    insta::assert_display_snapshot!(program);
}

#[test]
fn unary_expression2() {
    let program = vir_high! {
        procedure {
            locals {
                a: Bool,
            }
            bb1 {
                guard true && !a;
                assume label1 !a && !a;
                assume label1 !a && !a && !!!!a;
                goto { exit }
            }
        }
    };
    insta::assert_display_snapshot!(program);
    insta::assert_yaml_snapshot!(program);
}

#[test]
fn binary_operation() {
    let program = vir_high! {
        procedure {
            locals {
                a: Bool,
                b: Int,
            }
            bb1 {
                guard true;
                assume label1 a && a && a;
                assume label2 a || a || a;
                assume label3 a && a || a;
                assume label3 a || a && a;
                assume label4 b > 0 && a;
                assume label5 a && b > 0;
                goto { exit }
            }
        }
    };
    insta::assert_display_snapshot!(program);
    insta::assert_yaml_snapshot!(program);
}

#[test]
fn implies() {
    let program = vir_high! {
        procedure {
            locals {
                a: Bool,
            }
            bb1 {
                guard true -> false;
                assume label1 a && a && !a -> !a && a || a;
                assume label2 a && !a || a -> a;
                goto { exit }
            }
        }
    };
    insta::assert_display_snapshot!(program);
    insta::assert_yaml_snapshot!(program);
}

#[test]
fn quantifiers() {
    let program = vir_high! {
        procedure {
            locals {
                a: Bool,
            }
            bb1 {
                guard forall(|b: Int| a > 5 -> a > 4 && true, []);
                assume label1 forall(|a: Int| a > 5 -> a > 4 && true, [(a, true)]);
                goto { exit }
            }
        }
    };
    insta::assert_display_snapshot!(program);
    insta::assert_yaml_snapshot!(program);
}
