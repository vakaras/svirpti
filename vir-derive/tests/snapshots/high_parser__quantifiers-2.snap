---
source: vir-derive/tests/high_parser.rs
expression: program
---
sorts: []
axioms: []
functions: []
procedure:
  variables:
    - name: a
      sort: Bool
  basic_blocks:
    - label: entry
      guard:
        Constant:
          Bool: true
      statements: []
      successors:
        - 1
    - label: bb1
      guard:
        Quantifier:
          kind: ForAll
          variables:
            - name: b
              sort: Int
          triggers: []
          body:
            BinaryOperation:
              kind: Implies
              left:
                BinaryOperation:
                  kind: GtCmp
                  left:
                    Variable:
                      name: a
                  right:
                    Constant:
                      Int: 5
              right:
                BinaryOperation:
                  kind: And
                  left:
                    BinaryOperation:
                      kind: GtCmp
                      left:
                        Variable:
                          name: a
                      right:
                        Constant:
                          Int: 4
                  right:
                    Constant:
                      Bool: true
      statements:
        - Assume:
            assertion:
              Quantifier:
                kind: ForAll
                variables:
                  - name: a
                    sort: Int
                triggers:
                  - parts:
                      - Variable:
                          name: a
                      - Constant:
                          Bool: true
                body:
                  BinaryOperation:
                    kind: Implies
                    left:
                      BinaryOperation:
                        kind: GtCmp
                        left:
                          Variable:
                            name: a
                        right:
                          Constant:
                            Int: 5
                    right:
                      BinaryOperation:
                        kind: And
                        left:
                          BinaryOperation:
                            kind: GtCmp
                            left:
                              Variable:
                                name: a
                            right:
                              Constant:
                                Int: 4
                        right:
                          Constant:
                            Bool: true
            label: label1
      successors:
        - 2
    - label: exit
      guard:
        Constant:
          Bool: true
      statements: []
      successors: []
